source/extensions/common/wasm/wasm.cc

```diff
@@ -1016,6 +1037,153 @@ uint64_t Context::getCurrentTimeNanoseconds() {
       .count();
 }
 
+template <int N> struct marshall { char buf[N]; };
+
+absl::optional<std::string> serializeValue(Filters::Common::Expr::CelValue value) {
+  using Filters::Common::Expr::CelValue;
+  switch (value.type()) {
+  case CelValue::Type::kMessage:
+    if (value.MessageOrDie() != nullptr) {
+      std::string result;
+      value.MessageOrDie()->SerializeToString(&result);
+      return result;
+    }
+  case CelValue::Type::kString:
+    return std::string(value.StringOrDie().value());
+  case CelValue::Type::kBytes:
+    return std::string(value.BytesOrDie().value());
+  case CelValue::Type::kInt64: {
+    const size_t n = sizeof(int64_t);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.Int64OrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kUint64: {
+    const size_t n = sizeof(uint64_t);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.Uint64OrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kDouble: {
+    const size_t n = sizeof(double);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.DoubleOrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kBool: {
+    const size_t n = sizeof(bool);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.BoolOrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kDuration: {
+    const size_t n = sizeof(absl::Duration);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.DurationOrDie());
+    return std::string(out.buf, n);
+  }
+  case CelValue::Type::kTimestamp: {
+    const size_t n = sizeof(absl::Time);
+    marshall<n> out = absl::bit_cast<marshall<n>>(value.TimestampOrDie());
+    return std::string(out.buf, n);
+  }
+  default:
+    // TODO: lists and maps
+    break;
+  }
+
+  return {};
+}
+
+absl::optional<std::string> Context::getSelectorExpression(absl::string_view path) {
+  using Filters::Common::Expr::CelValue;
+  using google::api::expr::runtime::FieldBackedListImpl;
+  using google::api::expr::runtime::FieldBackedMapImpl;
+
+  bool first = true;
+  CelValue value;
+  Protobuf::Arena arena;
+  const StreamInfo::StreamInfo& info = decoder_callbacks_->streamInfo();
+  const auto request_headers = request_headers_ ? request_headers_ : access_log_request_headers_;
+  const auto response_headers =
+      response_headers_ ? response_headers_ : access_log_response_headers_;
+  const auto response_trailers =
+      response_trailers_ ? response_trailers_ : access_log_response_trailers_;
+
+  size_t start = 0;
+  while (true) {
+    if (start >= path.size()) {
+      break;
+    }
+
+    size_t end = path.find('\0', start);
+    if (end == absl::string_view::npos) {
```

Comment:
This should never be hit since we always null-terminate.  I suggest returning an error here.

Reply:
Changed to return immediately with empty result. Returning an error is rather painful since we don't have StatusOr.

Reply:
https://en.cppreference.com/w/cpp/utility/variant (when we get to c++17).

You can return std::pair<WasmResult, std::string> or return the WasmResult and the std::string indirectly.

Reply:
Done using output argument.